src/build/tenscript/build_phase.rs

```diff
@@ -59,30 +62,103 @@ impl BuildNode {
 
 #[derive(Debug)]
 enum Launch {
-    Seeded { face_alias: FaceAlias },
+    Scratch { face_alias: FaceAlias },
     NamedFace { face_alias: FaceAlias },
     IdentifiedFace { face_id: UniqueId },
 }
 
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Clone)]
+pub struct BaseAliases {
+    left_bot: FaceAlias,
+    left_top: FaceAlias,
+    right_bot: FaceAlias,
+    right_top: FaceAlias,
+    omni_left_bot: FaceAlias,
+    omni_right_bot: FaceAlias,
+}
+
+impl Default for BaseAliases {
+    fn default() -> Self {
+        Self {
+            left_bot: FaceAlias("Left::Bot".to_string()),
+            left_top: FaceAlias("Left::Top".to_string()),
+            right_bot: FaceAlias("Right::Bot".to_string()),
+            right_top: FaceAlias("Right::Top".to_string()),
+            omni_left_bot: FaceAlias("Omni::Left::Bot".to_string()),
+            omni_right_bot: FaceAlias("Omni::Right::Bot".to_string()),
+        }
+    }
+}
+
+impl BaseAliases {
+    pub fn spin_based(&self, spin: Spin) -> (&FaceAlias, &FaceAlias) {
+        match spin {
+            Spin::Left => (&self.left_bot, &self.left_top),
+            Spin::Right => (&self.right_bot, &self.right_top),
+        }
+    }
+
+    pub fn spin_double_based(&self, spin: Spin, needs_double: bool) -> &FaceAlias {
+        match spin {
+            Spin::Left if needs_double => &self.omni_left_bot,
+            Spin::Right if needs_double => &self.omni_right_bot,
+            Spin::Left => &self.left_bot,
+            Spin::Right => &self.right_bot,
+        }
+    }
+
+    pub fn not_top_of_single(&self, alias: &FaceAlias) -> bool {
+        !(&self.right_top == alias || &self.left_top == alias)
+    }
+
+    pub fn other_alias(&self, alias: &FaceAlias) -> &FaceAlias {
+        if alias == &self.right_bot {
+            &self.right_top
+        } else if alias == &self.left_bot {
+            &self.left_top
+        } else {
+            panic!("no other alias found")
+        }
+    }
+}
```

Comment:
Hmm, can't we make this just a bunch of static match statements? Do you ever want to change the base4 aliases at runtime?

Reply:
Go for it

Reply:
Nah.